node/src/keyshare/mod.rs

```diff
@@ -1,80 +1,396 @@
+///  - Temporary key storage is a collection of keyshares that are persisted as individual files.
+///    Each keyshare is identified by (epoch ID, domain ID, attempt ID).
+///
+/// Important: although "temporary key storage" uses the word "temporary", files must still be
+/// strongly persisted and it is not acceptable to delete the local files. The persistence
+/// requirements of the temporary and permanent key storages are the same (i.e. node operators
+/// must not lose them, for doing so is the same as permanently leaving the MPC network).
+///
+/// Keyshares persisted into temporary key storage are not guaranteed to be used; rather, the
+/// voting mechanism (vote_pk and vote_reshared) on the contract ultimately decides which
+/// keyshare to use for each domain (in case multiple attempts were made to generate or reshare
+/// the key).
+///
+/// Whenever the contract transitions into the Running state (which certifies the exact key to
+/// use for each domain), the keyshares are promoted from temporary key storage to permanent key
+/// storage. This is done by calling `load_keyset` when handling the Running state.
+///
+/// A subtle detail is that we may miss the Running state if it quickly transitions into
+/// another state (Initializing or Resharing). For the initializing state we do not actually need
+/// the existing keyshares, but for the resharing state we do. And that's not a problem:
+/// in the resharing state we also need to call `load_keyset`, and that will promote the keys into
+/// permanent storage if needed.
+pub struct KeyshareStorage {
+    temporary: TemporaryKeyStorage,
+    permanent: PermanentKeyStorage,
 }
 
-/// Factory to construct a KeyshareStorage implementation.
-pub enum KeyshareStorageFactory {
-    Gcp {
-        project_id: String,
-        secret_id: String,
-    },
-    Local {
-        home_dir: std::path::PathBuf,
-        encryption_key: [u8; 16],
-    },
-}
+impl KeyshareStorage {
+    /// Stores a single keyshare into temporary key storage. The keyshare's KeyEventId must be
+    /// unique. This must be called before sending in a vote_pk or vote_reshared.
+    pub async fn store_key(&self, key_share: Keyshare) -> anyhow::Result<()> {
+        self.temporary.store_keyshare(key_share).await
+    }
+
+    /// Before generating a key, we must call `ensure_can_generate_key` to check that we are able
+    /// to generate that key and use it afterwards. This requires:
+    ///  - The already generated keys exist either in permanent or temporary storage.
+    ///  - The current permanent key storage is either
+    ///    - In the same epoch as the key generation attempt, and whose keys are a prefix of the
+    ///      already generated keys.
+    ///    - In an older epoch. This can happen if we missed the previous transition from Resharing
+    ///      to Running before it transitions again into Initializing. This is fine.
```

Comment:
I have some trouble understanding this. Shouldn't we require that each node has a valid keyshare for each generated key in the contract? Not just a subset of those?

If the contract is in generating state and expects the nodes to have a key with a certain id, then a node not having that key seems problematic.

The epoch id mismatch due to missing a running state seems like it also should not happen: when we reshare a key, we index the new keyshare under the next epoch id, not the current epoch id. And if we transition from running to initializing, then the epoch id does not change.